File: src/gui/backends/sdl2/button.rs
use crate::gui::Position;
use crate::gui::Size;
use crate::gui::Rect;
use crate::gui::color::Color;

use super::Font;
use super::Widget;

pub struct Button<'a>
{
    label: String,
    raw: Widget<'a>,
}

impl<'a> Button<'_>
{
    const DEFAULT_SIZE: Size = Size(150, 40);
}

impl<'a> Button<'a>
{
    pub fn new(label: String, position: Position, color: Color, font: &'a Font<'a>) -> Button<'a>
    {
        let outer_part = Rect {
            color,
            size: Self::DEFAULT_SIZE,
            position,
        };

        // let (width, height) = font.raw.size_of(label.as_str()).unwrap();
        let surface = font.raw.render(label.as_str()).blended(font.color.rgba()).unwrap();

        Button {
            label,
            raw: Widget { raw: vec![outer_part], surfaces: vec![surface], font },
        }
    }
}

impl<'a> Button<'a>
{
    pub fn raw(&self) -> &Widget<'a> {
        &self.raw
    }

    pub fn from_raw(raw: Widget<'a>) -> Self where Self: Sized {
        Button {
            label: "".to_string(),
            raw: Widget::from_raw(raw),
        }
    }
}

#[cfg(test)]
mod test
{
    use crate::gui::color::Color;
    use crate::gui::window_flags::WindowFlags;
    use crate::gui::backends::sdl2::{Application, Window, WidgetType};

    use super::Button;

    #[test]
    fn test_button()
    {
        // let mut app = Application::new().unwrap();
        // let mut window = Window::new(&app, "test button", 800, 600, Color::WHITE, WindowFlags::Default);

        // let button = Button::new("label".to_string(), (10, 10).into(), (0, 255, 0).into());
        // window.add_button(&button);

        // app.draw_window(&mut window);
        // app.main_loop();
    }
}

File: src/gui/backends/sdl2/button.zig
const sdl2 = @import("sdl2.zig");

const std = @import("std");

const widget_mod = @import("widget.zig");
const font_mod = @import("font.zig");
const Size = @import("../../size.zig");
const Position = @import("../../position.zig");
const Rect = @import("../../rect.zig");
const Color = @import("../../color.zig");

pub const Button = @This();

label: []const u8,
raw_widget: widget_mod.Widget,

const DEFAULT_SIZE = Size { .width = 150, .height = 40 };
const Self = @This();

pub fn new(label: []const u8, position: Position, color: Color, font: *font_mod.Font, allocator: std.mem.Allocator) Button
{
    var outer_part = Rect{ .color = color, .size = DEFAULT_SIZE, .position = position };

    var rects = std.ArrayList(Rect).init(allocator);
    rects.append(outer_part) catch unreachable;

    var surface = font.draw_to_surface(label) catch unreachable;
    surface.rect.x = @intCast(c_int, position.x + (outer_part.size.width - @intCast(u32, surface.rect.w)) / 2);
    surface.rect.y = @intCast(c_int, position.y + (outer_part.size.height - @intCast(u32, surface.rect.h)) / 2);

    var surfaces = std.ArrayList(font_mod.TextSurface).init(allocator);
    surfaces.append(surface) catch unreachable;

    return Button {
        .label = label,
        .raw_widget = widget_mod.Widget.new(rects, surfaces)
    };
}

pub fn raw(self: *Self) *widget_mod.Widget
{
    return &self.*.raw_widget;
}

pub fn destroy(self: *Self) void
{
    self.raw_widget.destroy();
}

test "test button"
{

}